tls/extensions/s2n_server_renegotiation_info.c

```diff
@@ -27,28 +27,29 @@
 static bool s2n_renegotiation_info_should_send(struct s2n_connection *conn);
 static int s2n_renegotiation_info_send(struct s2n_connection *conn, struct s2n_stuffer *out);
 static int s2n_renegotiation_info_recv(struct s2n_connection *conn, struct s2n_stuffer *extension);
+static int s2n_renegotiation_info_if_missing(struct s2n_connection *conn);
 
 const s2n_extension_type s2n_server_renegotiation_info_extension = {
     .iana_value = TLS_EXTENSION_RENEGOTIATION_INFO,
-    .is_response = false,
     .send = s2n_renegotiation_info_send,
     .recv = s2n_renegotiation_info_recv,
     .should_send = s2n_renegotiation_info_should_send,
+    .if_missing = s2n_renegotiation_info_if_missing,
 
     /**
-     *= https://tools.ietf.org/rfc/rfc5746#3.4
-     *# *  If the extension is not present, the server does not support
-     *#    secure renegotiation; set secure_renegotiation flag to FALSE.
-     *#    In this case, some clients may want to terminate the handshake
-     *#    instead of continuing; see Section 4.1 for discussion.
-     *
-     * The conn->secure_renegotiation flag defaults to false, so this is a no-op.
-     * We do not terminate the handshake, although missing messaging for secure
-     * renegotiation degrades server security.
+     *= https://tools.ietf.org/rfc/rfc5746#3.6
+     *# Note that sending a "renegotiation_info" extension in response to a
+     *# ClientHello containing only the SCSV is an explicit exception to the
+     *# prohibition in RFC 5246, Section 7.4.1.4, on the server sending
+     *# unsolicited extensions and is only allowed because the client is
+     *# signaling its willingness to receive the extension via the
+     *# TLS_EMPTY_RENEGOTIATION_INFO_SCSV SCSV.
      *
-     * We could introduce an option to fail in this case in the future.
+     * This extension is technically a response extension, but doesn't
+     * follow any of the usual response extension rules.
+     * s2n-tls will therefore not treat it as a response extension.
      */
-    .if_missing = s2n_extension_noop_if_missing,
+    .is_response = false,
```

Comment:
what are the "usual response extension rules."

Reply:
The "is_response" flag turns on the logic for that restriction the compliance comment calls out: it ensures we only send the extension if the request version was received, and only accept the extension if the request version was sent.

We shouldn't enforce those rules here though, because this extension intentionally violates them.